cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1196 @@
+		if err != nil {
+			code := grpc.Code(err)
+			if code != grpcerror.OK && code != grpcerror.Canceled {
+				log.Warn("event feed receive from grpc stream failed",
+					zap.String("namespace", s.changefeed.Namespace),
+					zap.String("changefeed", s.changefeed.ID),
+					zap.Uint64("storeID", rs.storeID),
+					zap.String("addr", rs.storeAddr),
+					zap.Error(err))
+				return errors.Trace(err)
+			}
+			return context.Canceled
+		}
+
+		if len(cevent.Events) > 0 {
+			if err = s.sendRegionChangeEvents(ctx, cevent.Events, stream); err != nil {
+				return err
+			}
+		}
+		if cevent.ResolvedTs != nil {
+			s.metrics.batchResolvedSize.Observe(float64(len(cevent.ResolvedTs.Regions)))
+			if err = s.sendResolvedTs(ctx, cevent.ResolvedTs, stream); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (s *SharedClient) sendRegionChangeEvents(ctx context.Context, events []*cdcpb.Event, rs *requestedStream) error {
+	for _, event := range events {
+		regionID := event.RegionId
+		subscriptionID := SubscriptionID(event.RequestId)
+		if state := rs.getState(subscriptionID, regionID); state != nil {
+			sfEvent := newEventItem(event, state, rs)
+			slot := hashRegionID(regionID, len(s.workers))
+			if err := s.workers[slot].sendEvent(ctx, sfEvent); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *SharedClient) sendResolvedTs(ctx context.Context, resolvedTs *cdcpb.ResolvedTs, rs *requestedStream) error {
+	subscriptionID := SubscriptionID(resolvedTs.RequestId)
+	sfEvents := make([]statefulEvent, len(s.workers))
+	log.Debug("event feed get a ResolvedTs",
+		zap.String("namespace", s.changefeed.Namespace),
+		zap.String("changefeed", s.changefeed.ID),
+		zap.Any("subscriptionID", subscriptionID),
+		zap.Uint64("ResolvedTs", resolvedTs.Ts),
+		zap.Int("regionCount", len(resolvedTs.Regions)))
+
+	for _, regionID := range resolvedTs.Regions {
+		slot := hashRegionID(regionID, len(s.workers))
+		if sfEvents[slot].stream == nil {
+			sfEvents[slot] = newResolvedTsBatch(resolvedTs.Ts, rs)
+		}
+		x := &sfEvents[slot].resolvedTsBatch
+		if state := rs.getState(subscriptionID, regionID); state != nil {
+			x.regions = append(x.regions, state)
+		}
+	}
+
+	for i, sfEvent := range sfEvents {
+		if len(sfEvent.resolvedTsBatch.regions) > 0 {
+			sfEvent.stream = rs
+			if err := s.workers[i].sendEvent(ctx, sfEvent); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *requestedStore) appendRequest(sri singleRegionInfo) {
+	offset := r.nextStream.Add(1) % uint32(len(r.streams))
+	r.streams[offset].requests.In() <- sri
+}
+
+func (s *SharedClient) newStream(ctx context.Context, g *errgroup.Group, r *requestedStore) *requestedStream {
+	stream := &requestedStream{
+		streamID: streamIDGen.Add(1),
+		requests: chann.NewAutoDrainChann[singleRegionInfo](),
+	}
+
+	stream.requestedRegions.m = make(map[SubscriptionID]map[uint64]*regionFeedState)
+	stream.init = func(ctx context.Context) (err error) {
```

Comment:
init callback and preFetchForConnecting are unnecessary since stream could be initialized before runStream([L662](https://github.com/pingcap/tiflow/pull/9163/files#diff-76a50fd4045a2fd26a077fde7fa6f1e77ed9cce5871eb09a0588c74434a44c51R662)).

Reply:
`init` can be removed but seems `preFetchForConnecting` is still necessary.